Begin porting OpenIddictServerBuilderTests from 2.x.

Darthruneis opened the pull request:
- Some tests would not work without the changes to how the Options were retrieved.
- The skipped tests are those that are currently failing on all frameworks.
Will be continuing to try to fix those.
- The commented out tests have compiler errors.
Will need to go through and see which, if any, are still valid.

Related Work Items: #894

Posting this in a somewhat WIP state in case @kevinchalet has any feedback on the changes so far, and also to see what happens with the CI builds.

Darthruneis committed:
Begin porting OpenIdDictServerBuilderTests from 2.x.

- Some tests would not work without the changes to how the Options were retrieved.
- The skipped tests are those that are currently failing on all frameworks.
Will be continuing to try to fix those.
- The commented out tests have compiler errors.
Will need to go through and see which, if any, are still valid.

Related Work Items: #894

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddict.Server.Tests.csproj:
```diff
@@ -0,0 +1,37 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Moq" Version="4.13.1" />
```
Please use `$(MoqVersion)` to avoid hardcoding the Moq version.

Darthruneis replied:
Good to know, updated!

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddict.Server.Tests.csproj:
```diff
@@ -0,0 +1,37 @@
+
+  <ItemGroup>
+    <PackageReference Include="Moq" Version="4.13.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
```
This dependency is automagically added by Arcade and specifying it explicitly causes the exception seen in the CI logs.

Darthruneis replied:
Removed. I swore I had seen another of the Test projects including it, but looking again I can't. Maybe it was because it showed up in the NuGet Package Manager window in VS? Either way, removed.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddict.Server.Tests.csproj:
```diff
@@ -0,0 +1,37 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\OpenIddict.Abstractions\OpenIddict.Abstractions.csproj" />
```
This package is referenced by `OpenIddict.Server` so you should be able to remove it.

Darthruneis replied:
Yeah, good call. I added the Abstractions reference first while going through the 'fix all the red squigglies from namespaces that could not be found'. Didn't think about it being a transitive dependency.

Updated.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddict.Server.Tests.csproj:
```diff
@@ -0,0 +1,37 @@
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net472' ">
+    <Reference Include="System.Net.Http" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="3.1.0" />
```
Once you replace `3.1.0` by `$(ExtensionsVersion)`, you'll be able to remove most of these `ItemGroup`s.

Darthruneis replied:
Ah, nice. Updated!

kevinchalet reviewed (commented):
Thanks for your PR! I added a couple of notes that should help you with dependencies.

Darthruneis commented:
@kevinchalet One of the tests I was having trouble with were the certificate ones. What is the process used for generating the certificates used in the test projects? Is it just a self-signed cert created locally?

Reviewing your other comments now.

Darthruneis committed:
Handle comments on PR.

kevinchalet commented:
> @kevinchalet One of the tests I was having trouble with were the certificate ones. What is the process used for generating the certificates used in the test projects? Is it just a self-signed cert created locally?

The certificates stored as embedded resources, they are indeed self-signed certificates.

Assuming you're referring to the 2 skipped tests - that use certificate generation - I suspect that's because the `SUPPORTS_CERTIFICATE_GENERATION` compilation constant is not set. You'll need to set if for all TFMs except `net461` (e.g https://github.com/openiddict/openiddict-core/blob/dev/src/OpenIddict.Server/OpenIddict.Server.csproj#L33).

Darthruneis commented:
Yeah, I've got that line in there, but currently I don't have a certificate in the project. I was initially thinking I'd need to generate a new one, but I guess it should also be possible to just go back and get the one that was previously in the test project back in 2.x. That might be easier/less prone to new issues popping up in the test.

Darthruneis committed:
Add certificate and re-enable the certificate tests.

Darthruneis committed:
Update handler registration tests.

Remaining commented out tests are for methods that do not compile as the test is currently written.
Need to investigate whether these were truly removed or if maybe the namespace the extensions
come from are just not imported yet.

Darthruneis committed:
Replace some of the commented out 'Enable' tests with tests for the new methods.

Darthruneis committed:
Replace array initialization with Array.Empty to fix build.

Darthruneis committed:
Add more tests around exceptions.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs:
```diff
@@ -0,0 +1,2000 @@
+            Assert.Equal("services", exception.ParamName);
+        }
+
+        private class CustomContext : BaseContext
```
nit: can you please put them at the end of the file?

Darthruneis replied:
Done.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs:
```diff
@@ -0,0 +1,2000 @@
+
+        private class CustomContext : BaseContext
+        {
+            /// <summary>
```
You can remove the XML comments, they are not very useful in this case.

Darthruneis replied:
Resharper generated those when I used its 'generate missing members' functionality. I can remove them if you like, but it didn't take any time to add them either :).

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs:
```diff
@@ -0,0 +1,2000 @@
+            Action<OpenIddictServerHandlerDescriptor.Builder<CustomContext>> configuration = null;
+
+            // Act
+            Action action = () => builder.AddEventHandler<CustomContext>(configuration);
```
Technically, the exception is not thrown at this stage, so it's weird to put it in the "act" phase. Could you please combine the act and assert into a single "Act and assert" phase?

Darthruneis replied:
Updated all cases where this was being done in this way.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs:
```diff
@@ -0,0 +1,2000 @@
+            // Arrange
+            var services = CreateServices();
+            var builder = CreateBuilder(services);
+            Action<OpenIddictServerHandlerDescriptor.Builder<CustomContext>> configuration = null;
```
To simply things, feel free to remove that and do `builder.AddEventHandler<CustomContext>(configuration: null)` like in the other tests.

Darthruneis replied:
I used the variable so that I could use `nameof(configuration)` in the assertion, rather than hardcoding the string value for the assert. I avoid hardcoding strings when possible - do you think there's a negative to using `nameof` in this way in the context of a unit test?

kevinchalet replied:
It's certainly not a bad practice, it's just that I haven't used this approach anywhere else, so it's a bit inconsistent. But I'm fine keeping it this way if you like it.

kevinchalet commented on test/OpenIddict.Server.Tests/OpenIddictServerBuilderTests.cs:
```diff
@@ -0,0 +1,2000 @@
+            Assert.Equal(nameof(key), exception.ParamName);
+        }
+
+        private class FakeAsymmetricSecurityKey : AsymmetricSecurityKey
```
May I suggest using Moq here? I used it in the ASOS tests and it worked fine in this case: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/dev/test/AspNet.Security.OpenIdConnect.Server.Tests/OpenIdConnectServerExtensionsTests.cs#L530

Darthruneis replied:
Verified tests still pass using Moq, so replaced with Moq.

kevinchalet reviewed (commented):
It looks excellent, thanks a lot!

Do you need help with the remaining tests?

Darthruneis committed:
Handle PR comments.

Darthruneis commented:
> Do you need help with the remaining tests?

I think I've got most of the coverage of the ServerBuilder class now. Anything related to the certificates or encryption may need to be built out a bit more. I've shied away those due to a lack of experience working with certs/encryption.

Regarding the other classes which are referenced in #894 , I intend to start on those next. Do you have any preference between:
- A different PR for each of the class that is being tested, or 
- a single PR (this one) to encompass all of 894?

kevinchalet commented:
FYI, the fact you're using `action` delegate variables results in VS adding many "info" messages suggesting replacing them by local functions:

![image](https://user-images.githubusercontent.com/6998306/84661601-df626700-af1a-11ea-8300-e1bfc089b1e5.png)

This is not blocking for this PR, but to reduce the noise in VS, you'll want probably want to avoid the intermediate variable and do `Assert.Throws<ArgumentNullException>(() => builder.AddEncryptionKey(key))`.

> I think I've got most of the coverage of the ServerBuilder class now. Anything related to the certificates or encryption may need to be built out a bit more. I've shied away those due to a lack of experience working with certs/encryption.

Great. I'll take a look once this PR is merged.

> Regarding the other classes which are referenced in #894 , I intend to start on those next. Do you have any preference between.

No preference. Feel free to opt for the one who better suits your flow. If you opt for the second option, please let me know and I'll merge this PR today.

Thanks!

Darthruneis committed:
Clean up info messages.

Darthruneis commented:
Updated to clean up those info messages. I've had info messages turned off in VS for ages so I had no idea it was polluting those up so much.

As for the PR, I prefer to merge it in then. Easier to make a new branch then and work on the next set of tests that need ported, and it allows the tests to be in place as the safety net they hopefully will act as.

kevinchalet commented:
PR merged, thanks a lot for your contribution! 👏
